File: fontes/br.com.wdc.shopping/br.com.wdc.shopping.view.teavm/teavm.web/src/main/java/br/com/wdc/shopping/view/teavm/infra/EncryptedLocalStorage.java
package br.com.wdc.shopping.view.teavm.infra;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.Map;

import org.teavm.jso.JSBody;
import org.teavm.jso.JSObject;

import br.com.wdc.framework.commons.storage.ClientStorage;
import br.com.wdc.shopping.view.teavm.commons.interop.JsObjectConsumer;
import br.com.wdc.shopping.view.teavm.commons.interop.JsStringConsumer;

/**
 * Implementação de {@link ClientStorage} com criptografia AES-GCM via Web Crypto API.
 * <p>
 * A chave AES-256 é criada com {@code extractable: false} e persiste no IndexedDB
 * da origem. Por ser não-exportável, os bytes da chave nunca ficam acessíveis ao
 * JavaScript — nem por XSS. Os valores são cifrados com IV aleatório por entrada
 * e armazenados em {@code localStorage} no formato {@code base64(iv[12] + ciphertext)}.
 * <p>
 * Fluxo de inicialização:
 * <ol>
 *   <li>Abre (ou cria) IndexedDB {@code _sec}, object store {@code k}.</li>
 *   <li>Se a chave já existe (reloads subsequentes): recupera e usa.</li>
 *   <li>Se não existe (primeira execução): gera com
 *       {@code crypto.subtle.generateKey}, armazena no IndexedDB.</li>
 *   <li>Decifra todas as entradas {@code sec.*} do localStorage para o cache em memória.</li>
 *   <li>Chama {@code onReady} — a partir daí, leituras e escritas são síncronas
 *       (via cache) com persistência assíncrona.</li>
 * </ol>
 * <p>
 * Fallback: se o IndexedDB não estiver disponível (ex.: modo privado em alguns browsers),
 * o storage funciona somente em memória (valores não sobrevivem ao reload).
 * <p>
 * <strong>Deve-se chamar {@link #initialize(Runnable)} antes de usar o storage.</strong>
 */
public class EncryptedLocalStorage implements ClientStorage {

    public static final EncryptedLocalStorage INSTANCE = new EncryptedLocalStorage();

    /** Cache em memória dos valores decifrados. Populado durante {@link #initialize}. */
    private final Map<String, String> cache = new LinkedHashMap<>();

    /** Chave AES-GCM não-exportável, obtida do IndexedDB. Null se indisponível. */
    private static JSObject cryptoKey;

    /**
     * Prefixo no localStorage: {@code "{shellId}:sec."}.
     * Configurado por {@link #configure(String)} antes de {@link #initialize(Runnable)}.
     */
    private static String LS_PREFIX = "sec.";

    /**
     * Configura o prefixo de namespace específico do shell.
     * Deve ser chamado antes de {@link #initialize(Runnable)}.
     *
     * @param shellId identificador curto do shell (ex: {@code "tw"})
     */
    public static void configure(String shellId) {
        LS_PREFIX = shellId + ":sec.";
    }

    private EncryptedLocalStorage() {
    }

    /**
     * Inicializa a chave de criptografia e pré-decifra todos os valores armazenados
     * para o cache em memória. Deve ser chamado uma vez antes de {@code app.start()}.
     *
     * @param onReady callback chamado quando o storage está pronto para uso
     */
    public static void initialize(Runnable onReady) {
        openOrCreateKey(key -> {
            cryptoKey = key;
            if (key == null) {
                // IndexedDB indisponível — funciona apenas em memória
                onReady.run();
                return;
            }

            // Coleta snapshot síncrono das chaves sec.* no localStorage
            int len = lsLength();
            var secEntries = new ArrayList<String>();
            for (int i = 0; i < len; i++) {
                String rawKey = lsKey(i);
                if (rawKey != null && rawKey.startsWith(LS_PREFIX)) {
                    secEntries.add(rawKey);
                }
            }

            if (secEntries.isEmpty()) {
                onReady.run();
                return;
            }

            int[] remaining = { secEntries.size() };

            for (String rawKey : secEntries) {
                String shortKey = rawKey.substring(LS_PREFIX.length());
                String b64 = lsGetItem(rawKey);
                if (b64 == null || b64.isEmpty()) {
                    if (--remaining[0] == 0) onReady.run();
                    continue;
                }
                decryptValue(key, b64, plaintext -> {
                    if (plaintext != null && !plaintext.isEmpty()) {
                        INSTANCE.cache.put(shortKey, plaintext);
                    }
                    // Else: valor com chave antiga/corrompido — descartado silenciosamente
                    if (--remaining[0] == 0) onReady.run();
                });
            }
        });
    }

    @Override
    public ClientStorage secure() {
        return this;
    }

    @Override
    public String get(String key) {
        return cache.get(key);
    }

    @Override
    public void set(String key, String value) {
        if (value == null) {
            remove(key);
            return;
        }
        cache.put(key, value);
        if (cryptoKey != null) {
            // Persiste assincronamente no localStorage cifrado.
            // Verifica cache.containsKey ao final para evitar race condition com remove():
            // se remove() foi chamado antes do callback completar, não grava no localStorage.
            encryptValue(cryptoKey, value, b64 -> {
                if (b64 != null && !b64.isEmpty() && cache.containsKey(key)) {
                    lsSetItem(LS_PREFIX + key, b64);
                }
            });
        }
    }

    @Override
    public void remove(String key) {
        cache.remove(key);
        lsRemoveItem(LS_PREFIX + key);
    }

    @Override
    public Map<String, String> all() {
        return new LinkedHashMap<>(cache);
    }

    // -- JS interop --

    /**
     * Abre (ou cria) o IndexedDB {@code _sec} e retorna a chave AES-GCM armazenada.
     * Se não existir, gera uma nova chave não-exportável e a persiste no IndexedDB.
     * Chama callback com {@code null} se o IndexedDB não estiver disponível.
     */
    @JSBody(params = { "callback" }, script = """
            try {
                var req = indexedDB.open('_sec', 1);
                req.onupgradeneeded = function(e) {
                    e.target.result.createObjectStore('k');
                };
                req.onsuccess = function(e) {
                    var db = e.target.result;
                    var tx = db.transaction('k', 'readwrite');
                    var store = tx.objectStore('k');
                    var get = store.get(0);
                    get.onsuccess = function() {
                        if (get.result) {
                            callback(get.result);
                        } else {
                            crypto.subtle.generateKey(
                                { name: 'AES-GCM', length: 256 },
                                false,
                                ['encrypt', 'decrypt']
                            ).then(function(key) {
                                var tx2 = db.transaction('k', 'readwrite');
                                tx2.objectStore('k').put(key, 0);
                                callback(key);
                            }).catch(function(err) {
                                console.error('[EncryptedStorage] key generation failed:', err);
                                callback(null);
                            });
                        }
                    };
                    get.onerror = function() { callback(null); };
                };
                req.onerror = function(e) {
                    console.warn('[EncryptedStorage] IndexedDB open failed:', e.target.error);
                    callback(null);
                };
            } catch(e) {
                console.warn('[EncryptedStorage] IndexedDB unavailable:', e);
                callback(null);
            }
            """)
    private static native void openOrCreateKey(JsObjectConsumer callback);

    /**
     * Cifra {@code text} com AES-GCM usando a chave fornecida.
     * Retorna {@code base64(iv[12] + ciphertext)} via callback, ou string vazia em caso de erro.
     */
    @JSBody(params = { "key", "text", "callback" }, script = """
            var iv = crypto.getRandomValues(new Uint8Array(12));
            crypto.subtle.encrypt({ name: 'AES-GCM', iv: iv }, key,
                    new TextEncoder().encode(text))
            .then(function(ct) {
                var b = new Uint8Array(12 + ct.byteLength);
                b.set(iv, 0);
                b.set(new Uint8Array(ct), 12);
                var s = '';
                for (var i = 0; i < b.length; i++) s += String.fromCharCode(b[i]);
                callback(btoa(s));
            }).catch(function(err) {
                console.error('[EncryptedStorage] encrypt failed:', err);
                callback('');
            });
            """)
    private static native void encryptValue(JSObject key, String text, JsStringConsumer callback);

    /**
     * Decifra um valor {@code base64(iv[12] + ciphertext)} produzido por {@link #encryptValue}.
     * Retorna o texto plano via callback, ou string vazia se falhar
     * (chave diferente, dado corrompido, etc.).
     */
    @JSBody(params = { "key", "b64", "callback" }, script = """
            try {
                var s = atob(b64);
                var b = new Uint8Array(s.length);
                for (var i = 0; i < s.length; i++) b[i] = s.charCodeAt(i);
                crypto.subtle.decrypt({ name: 'AES-GCM', iv: b.slice(0, 12) },
                        key, b.slice(12))
                .then(function(pt) {
                    callback(new TextDecoder().decode(new Uint8Array(pt)));
                }).catch(function(err) {
                    console.warn('[EncryptedStorage] decrypt failed (stale key or corrupted data):', err);
                    callback('');
                });
            } catch(e) {
                callback('');
            }
            """)
    private static native void decryptValue(JSObject key, String b64, JsStringConsumer callback);

    @JSBody(params = {}, script = "try { return localStorage.length; } catch(e) { return 0; }")
    private static native int lsLength();

    @JSBody(params = { "i" }, script = "try { return localStorage.key(i); } catch(e) { return null; }")
    private static native String lsKey(int i);

    @JSBody(params = { "key" }, script = "try { return localStorage.getItem(key); } catch(e) { return null; }")
    private static native String lsGetItem(String key);

    @JSBody(params = { "key", "val" }, script = "try { localStorage.setItem(key, val); } catch(e) {}")
    private static native void lsSetItem(String key, String val);

    @JSBody(params = { "key" }, script = "try { localStorage.removeItem(key); } catch(e) {}")
    private static native void lsRemoveItem(String key);
}

File: fontes/br.com.wdc.framework/br.com.wdc.framework.cube.remote/remote.bridge.react/src/bridge/EncryptedLocalStorage.ts
/**
 * AES-GCM encrypted localStorage implementation.
 *
 * The AES-256 key is created with `extractable: false` and persisted in the
 * origin's IndexedDB (`_sec` database, object store `k`). Because the key is
 * non-extractable, its raw bytes are never accessible to JavaScript — not even
 * via `crypto.subtle.exportKey()` — which limits the impact of an XSS attack.
 *
 * Values are stored in `localStorage` under a `sec.` prefix as
 * `base64(iv[12] || ciphertext)`. A fresh random IV is generated per write.
 *
 * ### Lifecycle
 * 1. Call `EncryptedLocalStorage.initialize()` once before the app starts.
 * 2. It opens (or creates) the IndexedDB key, then decrypts all existing
 *    `sec.*` entries into an in-memory cache.
 * 3. After the returned Promise resolves, `get`/`set`/`remove` are synchronous
 *    (cache-backed). Writes persist asynchronously (fire-and-forget).
 *
 * ### Fallback
 * If IndexedDB is unavailable (e.g. some private-browsing modes), the storage
 * works in-memory only — data is not persisted across reloads.
 */

import { ClientStorage } from './ClientStorage'

const IDB_NAME = '_sec'
const IDB_STORE = 'k'
const IDB_KEY = 0

export class EncryptedLocalStorage implements ClientStorage {
  private readonly cache = new Map<string, string>()
  private cryptoKey: CryptoKey | null = null
  private readonly lsPrefix: string
  private readonly storage: Storage

  /**
   * @param syncNamespace key namespace that qualifies entries for WebSocket sync;
   *                      convention: `'~<shellId>:'` (e.g. `'~rr:'`).
   *                      Prepended to every stored key and stripped in `all()`.
   *                      Also used to derive the storage prefix: `'<syncNamespace>sec.'`.
   * @param storage       backing Web Storage (`localStorage` or `sessionStorage`).
   *                      Defaults to `localStorage`.
   */
  constructor(private readonly syncNamespace = '', storage?: Storage) {
    this.lsPrefix = `${syncNamespace}sec.`
    this.storage = storage ?? localStorage
  }

  get secure(): ClientStorage {
    return this
  }

  get(key: string): string | null {
    return this.cache.get(key) ?? null
  }

  set(key: string, value: string): void {
    this.cache.set(key, value)
    if (this.cryptoKey) {
      this.encryptAndStore(key, value, this.cryptoKey).catch(() => {
        // fire-and-forget; failures are non-fatal
      })
    }
  }

  remove(key: string): void {
    this.cache.delete(key)
    this.storage.removeItem(this.lsPrefix + key)
  }

  all(): Record<string, string> {
    const result: Record<string, string> = {}
    for (const [k, v] of this.cache) {
      result[k] = v
    }
    return result
  }

  /**
   * Initialises the AES-GCM key from IndexedDB and pre-decrypts all stored
   * `sec.*` entries into the in-memory cache. Must be awaited before the app
   * starts so that `get()` returns correct values synchronously.
   */
  async initialize(): Promise<void> {
    try {
      this.cryptoKey = await openOrCreateKey()
    } catch {
      // IndexedDB unavailable — in-memory only
      return
    }

    if (!this.cryptoKey) return

    const key = this.cryptoKey
    const len = this.storage.length
    const entries: Array<{ short: string; b64: string }> = []
    for (let i = 0; i < len; i++) {
      const rawKey = this.storage.key(i)
      if (rawKey && rawKey.startsWith(this.lsPrefix)) {
        const b64 = this.storage.getItem(rawKey)
        if (b64) entries.push({ short: rawKey.slice(this.lsPrefix.length), b64 })
      }
    }

    await Promise.all(
      entries.map(async ({ short, b64 }) => {
        const plaintext = await decryptValue(key, b64)
        if (plaintext !== null) {
          this.cache.set(short, plaintext)
        }
        // null means stale key or corrupted data — silently discard
      }),
    )
  }

  private async encryptAndStore(key: string, value: string, cryptoKey: CryptoKey): Promise<void> {
    const b64 = await encryptValue(cryptoKey, value)
    this.storage.setItem(this.lsPrefix + key, b64)
  }
}

// ---------------------------------------------------------------------------
// IndexedDB helpers
// ---------------------------------------------------------------------------

/** Opens (or creates) the `_sec` IndexedDB and returns the stored AES key,
 *  generating and persisting a new one if none exists yet. */
function openOrCreateKey(): Promise<CryptoKey> {
  return new Promise((resolve, reject) => {
    const req = indexedDB.open(IDB_NAME, 1)

    req.onupgradeneeded = (e) => {
      ;(e.target as IDBOpenDBRequest).result.createObjectStore(IDB_STORE)
    }

    req.onerror = () => reject(req.error)

    req.onsuccess = () => {
      const db = req.result
      const tx = db.transaction(IDB_STORE, 'readwrite')
      const store = tx.objectStore(IDB_STORE)
      const get = store.get(IDB_KEY)

      get.onsuccess = () => {
        if (get.result) {
          resolve(get.result as CryptoKey)
          return
        }
        // No key yet — generate one
        crypto.subtle
          .generateKey({ name: 'AES-GCM', length: 256 }, false, ['encrypt', 'decrypt'])
          .then((key) => {
            const tx2 = db.transaction(IDB_STORE, 'readwrite')
            tx2.objectStore(IDB_STORE).put(key, IDB_KEY)
            resolve(key)
          })
          .catch(reject)
      }

      get.onerror = () => reject(get.error)
    }
  })
}

// ---------------------------------------------------------------------------
// Crypto helpers
// ---------------------------------------------------------------------------

/** Encrypts `text` and returns `base64(iv[12] || ciphertext)`. */
async function encryptValue(key: CryptoKey, text: string): Promise<string> {
  const iv = crypto.getRandomValues(new Uint8Array(12))
  const encoded = new TextEncoder().encode(text)
  const ct = await crypto.subtle.encrypt({ name: 'AES-GCM', iv }, key, encoded)
  const combined = new Uint8Array(12 + ct.byteLength)
  combined.set(iv, 0)
  combined.set(new Uint8Array(ct), 12)
  let s = ''
  for (let i = 0; i < combined.length; i++) s += String.fromCharCode(combined[i])
  return btoa(s)
}

/** Decrypts a `base64(iv[12] || ciphertext)` value. Returns `null` on failure. */
async function decryptValue(key: CryptoKey, b64: string): Promise<string | null> {
  try {
    const s = atob(b64)
    const b = new Uint8Array(s.length)
    for (let i = 0; i < s.length; i++) b[i] = s.charCodeAt(i)
    const pt = await crypto.subtle.decrypt({ name: 'AES-GCM', iv: b.slice(0, 12) }, key, b.slice(12))
    return new TextDecoder().decode(new Uint8Array(pt))
  } catch {
    return null
  }
}
